### PR TITLE
Distribute commands to subset of partitions

### LINKED ISSUE #16385: Extend generalized command distribution for subsets of partitions
This is needed for:
- Process Instance Migration of Message Subscriptions: reliably communicate the `MessageSubscription:MIGRATE` command 
- Replacing the reliable communication of Message Subscription Opening, Closing, and Correlation to benefit from generalized command distribution, like exponential backoff, actor yielding (not yet supported), caching (not yet supported), ordered distribution (not yet supported), metrics (not yet supported), etc

### PR DESCRIPTION
**Pull request by korthout**
## Description

Adds the ability to distribute a command to a subset of partitions rather than to all partitions.

This PR is mostly a bunch of refactorings that I believe improve the understandability of the concept and code.

I've also introduced a FakeProcessingResultBuilder that we can use when writing unit tests for parts of the engine. While we typically favor integration-style tests using the APIs available to the users, we sometimes have a need for testing something in isolation. This result builder fits really well here to test the command distribution behavior class.

## Related issues

closes #16385

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/distribution/CommandDistributionBehavior.java [hunk 1]
```diff
@@ -45,20 +60,41 @@ public CommandDistributionBehavior(
...
   public <T extends UnifiedRecordValue> void distributeCommand(
       final long distributionKey, final TypedRecord<T> command) {
-    if (otherPartitions.isEmpty()) {
+    distributeCommand(distributionKey, command, otherPartitions);
+  }
+
+  /**
+   * Distributes a command to the specified partitions.
+   *
+   * @param distributionKey the key to identify this unique command distribution. The key is used to
```
Comment (berkaycanbc) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/distribution/CommandDistributionBehavior.java:81]: 🔧 The second sentence can be re-written to make it more understandable. E.g.

`The key used for three purposes: storing the pending distribution as the key of distributed command, identifying for the distributed command's partition, and correlating the ACKNOWLEDGE command to the pending distribution.`

Or maybe something similar.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/distribution/CommandDistributionBehavior.java [hunk 2]
```diff
@@ -45,20 +60,41 @@ public CommandDistributionBehavior(
...
+   */
+  public <T extends UnifiedRecordValue> void distributeCommand(
+      final long distributionKey, final TypedRecord<T> command, final List<Integer> partitions) {
+    if (partitions.isEmpty()) {
       return;
     }
 
     final var distributionRecord =
-        new CommandDistributionRecord()
+        commandDistributionStarted
```
Comment (berkaycanbc) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/distribution/CommandDistributionBehavior.java:97]: 💭 What about implementing a reset method for the record and calling it before or after each usage? For now, it is not used around, so seems clear but for future changes, that can make the re-using more explicit. WDYT?

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/distribution/CommandDistributionBehavior.java [hunk 3]
```diff
@@ -67,49 +103,57 @@ public <T extends UnifiedRecordValue> void distributeCommand(
...
-  public <T extends UnifiedRecordValue> void acknowledgeCommand(
-      final long distributionKey, final TypedRecord<T> command) {
+  /**
+   * Acknowledges that a command was distributed to another partition successfully.
+   *
+   * @param command the command that was distributed
+   */
+  public <T extends UnifiedRecordValue> void acknowledgeCommand(final TypedRecord<T> command) {
+    final long distributionKey = command.getKey();
     final var distributionRecord =
```
Comment (berkaycanbc) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/distribution/CommandDistributionBehavior.java:150]: ❓Why we didn't create a re-usable record for the acknowledged command as we did refactoring for others?

**Review by berkaycanbc — CHANGES_REQUESTED** [review]
Thanks @korthout! 🚀 

⭐ I really like the idea of testing this in isolation.

👀 I've added some comments and suggestions, please take a look.

**Comment by Zelldon** [comment]
@berkaycanbc @korthout what is the status of this?

**Comment by korthout** [comment]
@Zelldon it's part of process instance migration, but the functionality is not needed right away. Focus was elsewhere for the time being. Please keep this open for when we need it. Until then, there's no need to merge this as it will not be in use.

**Comment by korthout** [comment]
This is needed for [PIM phase 3](https://github.com/camunda/zeebe/issues/17349) as part of:
- https://github.com/camunda/zeebe/issues/16386

**Comment by korthout** [comment]
@berkaycanbc I've rebased on top of `main` to allow running the new SDK test summary. The only commits that are new are `9ecfa38..e44d0eb`

**Review by berkaycanbc — APPROVED** [review]
Thanks for the latest changes @korthout, LGTM 🚀.

**Pull request merged**